library/core/src/str/validations.rs

```diff
@@ -105,21 +106,36 @@ const NONASCII_MASK: usize = 0x80808080_80808080u64 as usize;
 
 /// Returns `true` if any byte in the word `x` is nonascii (>= 128).
 #[inline]
-fn contains_nonascii(x: usize) -> bool {
+const fn contains_nonascii(x: usize) -> bool {
     (x & NONASCII_MASK) != 0
 }
 
 /// Walks through `v` checking that it's a valid UTF-8 sequence,
 /// returning `Ok(())` in that case, or, if it is invalid, `Err(err)`.
 #[inline(always)]
-pub(super) fn run_utf8_validation(v: &[u8]) -> Result<(), Utf8Error> {
+#[rustc_const_unstable(feature = "str_internals", issue = "none")]
+pub(super) const fn run_utf8_validation(v: &[u8]) -> Result<(), Utf8Error> {
     let mut index = 0;
     let len = v.len();
 
     let usize_bytes = mem::size_of::<usize>();
     let ascii_block_size = 2 * usize_bytes;
     let blocks_end = if len >= ascii_block_size { len - ascii_block_size + 1 } else { 0 };
-    let align = v.as_ptr().align_offset(usize_bytes);
+
+    fn rt_align(v: &[u8], usize_bytes: usize) -> usize {
+        v.as_ptr().align_offset(usize_bytes)
+    }
+
+    const fn ctfe_align(_: &[u8], _: usize) -> usize {
+        usize::MAX
+    }
+
+    // SAFETY:
+    // It is permisseble for `align_offset` to always return `usize::MAX`,
+    // algorithm correctness can not depend on `align_offset` returning non-max values.
+    //
+    // As such the behaviour can't change after replacing `align_offset` with `usize::MAX`, only performance can.
+    let align = unsafe { const_eval_select((v, usize_bytes), ctfe_align, rt_align) };
```

Comment:
It seems better to put this inside align_offset itself, since the justification should work for all calls to it, right?

Reply:
While I think it would be great to put this inside `align_offset` itself, wouldn't this allow for user code to behave differently in `const` and non-`const` context?
```rust
const fn is_likely_called_in_ctfe() -> bool {
    let arr = [0u8; 2];
    (&arr[0] as *const u8).align_offset(2) == usize::MAX 
        && (&arr[1] as *const u8).align_offset(2) == usize::MAX
}
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1b4b921b3efb38003e3d740accb8d48a))

Reply:
Actually even `(&0u8 as *const u8).align_offset(1) == usize::MAX` will do. Basically while `const_eval_select` in the caller of  `align_offset` shouldn't change behaviour, `const_eval_select` in `align_offset` itself clearly changes it's return value depending on whatever it's called at runtime or compile time.

Sadly, it seems that we don't want to allow this, right? :(

Reply:
> `const_eval_select` in `align_offset` itself clearly changes it's return value depending on whatever it's called at runtime or compile time.
> 
> Sadly, it seems that we don't want to allow this, right? :(

Actually, [the docs](https://doc.rust-lang.org/std/primitive.pointer.html#method.align_offset) specifically state

> If it is not possible to align the pointer, the implementation returns usize::MAX. It is permissible for the implementation to always return usize::MAX. Only your algorithm’s performance can depend on getting a usable offset here, not its correctness.
> There are no guarantees whatsoever that offsetting the pointer will not overflow or go beyond the allocation that the pointer points into. It is up to the caller to ensure that the returned offset is correct in all terms other than alignment.

The only reason `align_offset` even exists is because I wanted to enable miri to do such trickery back when it wasn't able to do ptr2int conversions.

So, we can definitely make `align_offset` behave this way in const fn, as long as we have a tracking issue for the const stabilization that specifically raises this concern before we bake it into a stabilized function.

Reply:
I can make a follow-up PR then

Reply:
I have a slight preference for doing a PR for just `align_offset` first, since then this PR does less work, but if you prefer to first merge this one I'm ok with that.

Reply:
I can do either way if you prefer `align_offset` first, then so be it :)